File: app/src/main/java/com/app/ralaunch/game/AssemblyPatcher.java
package com.app.ralaunch.game;

import android.content.Context;

import com.app.ralaunch.utils.AppLogger;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 程序集补丁工具
 * 
 * 从已解压的 MonoMod 目录加载补丁程序集，并替换游戏目录中的对应文件
 * 
 * @author RA Launcher Team
 */
public class AssemblyPatcher {
    private static final String TAG = "AssemblyPatcher";
    private static final String MONOMOD_DIR = "MonoMod";

    /**
     * 应用 MonoMod 补丁到游戏目录
     *
     * @param context Android上下文
     * @param gameDirectory 游戏目录路径
     * @return 替换的程序集数量，失败返回 -1
     */
    public static int applyMonoModPatches(Context context, String gameDirectory) {
        return applyMonoModPatches(context, gameDirectory, true);
    }

    /**
     * 应用 MonoMod 补丁到游戏目录（可控制日志级别）
     *
     * @param context Android上下文
     * @param gameDirectory 游戏目录路径
     * @param verboseLog 是否输出详细日志
     * @return 替换的程序集数量，失败返回 -1
     */
    public static int applyMonoModPatches(Context context, String gameDirectory, boolean verboseLog) {
        try {
            Map<String, byte[]> patchAssemblies = loadPatchArchive(context);
            if (patchAssemblies.isEmpty()) {
                return 0;
            }

            // 2. 扫描游戏目录中的程序集
            File gameDir = new File(gameDirectory);
            List<File> gameAssemblies = findGameAssemblies(gameDir);

            int patchedCount = 0;
            for (File assemblyFile : gameAssemblies) {
                String assemblyName = assemblyFile.getName();

                if (patchAssemblies.containsKey(assemblyName)) {
                    if (replaceAssembly(assemblyFile, patchAssemblies.get(assemblyName))) {
                        patchedCount++;
                    }
                }
            }
            return patchedCount;

        } catch (Exception e) {
            AppLogger.error(TAG, "应用补丁失败", e);
            return -1;
        }
    }

    /**
     * 从解压后的 MonoMod 目录加载补丁程序集
     */
    private static Map<String, byte[]> loadPatchArchive(Context context) {
        Map<String, byte[]> assemblies = new HashMap<>();

        try {
            File monoModDir = new File(context.getFilesDir(), MONOMOD_DIR);

            if (!monoModDir.exists() || !monoModDir.isDirectory()) {
                return assemblies;
            }

            List<File> dllFiles = findDllFiles(monoModDir);

            for (File dllFile : dllFiles) {
                try {
                    byte[] assemblyData = java.nio.file.Files.readAllBytes(dllFile.toPath());
                    String fileName = dllFile.getName();
                    assemblies.put(fileName, assemblyData);
                } catch (IOException e) {
                }
            }

        } catch (Exception e) {
            AppLogger.error(TAG, "加载 MonoMod 补丁失败", e);
        }

        return assemblies;
    }

    /**
     * 递归查找目录中的所有 .dll 文件
     */
    private static List<File> findDllFiles(File directory) {
        List<File> dllFiles = new ArrayList<>();

        File[] files = directory.listFiles();
        if (files != null) {
            for (File file : files) {
                if (file.isDirectory()) {
                    dllFiles.addAll(findDllFiles(file));
                } else if (file.getName().endsWith(".dll")) {
                    dllFiles.add(file);
                }
            }
        }

        return dllFiles;
    }

    /**
     * 扫描游戏目录，查找所有 .dll 程序集
     */
    private static List<File> findGameAssemblies(File directory) {
        List<File> assemblies = new ArrayList<>();

        if (!directory.exists() || !directory.isDirectory()) {
            return assemblies;
        }

        File[] files = directory.listFiles();
        if (files == null) {
            return assemblies;
        }

        for (File file : files) {
            if (file.isDirectory()) {
                assemblies.addAll(findGameAssemblies(file));
            } else if (file.getName().endsWith(".dll")) {
                assemblies.add(file);
            }
        }

        return assemblies;
    }

    /**
     * 替换程序集文件
     */
    private static boolean replaceAssembly(File targetFile, byte[] assemblyData) {
        try {
            FileOutputStream outputStream = new FileOutputStream(targetFile);
            outputStream.write(assemblyData);
            outputStream.close();
            return true;
        } catch (IOException e) {
            AppLogger.error(TAG, "替换失败: " + targetFile.getName(), e);
            return false;
        }
    }
}

File: app/src/main/java/com/app/ralaunch/core/platform/runtime/AssemblyPatcher.kt
package com.app.ralaunch.core.platform.runtime

import android.content.Context
import timber.log.Timber
import com.app.ralaunch.core.extractor.ArchiveExtractor
import com.app.ralaunch.core.common.util.TemporaryFileAcquirer
import org.koin.java.KoinJavaComponent
import java.io.File
import java.io.FileOutputStream
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption

/**
 * 程序集补丁工具
 */
object AssemblyPatcher {
    const val MONOMOD_DIR = "monomod"
    private const val ASSETS_MONOMOD_ZIP = "MonoMod.zip"

    @JvmStatic
    fun getMonoModInstallPath(): Path {
        val context: Context = KoinJavaComponent.get(Context::class.java)
        val externalFilesDir = context.getExternalFilesDir(null)
        return Paths.get(externalFilesDir?.absolutePath ?: "", MONOMOD_DIR)
    }

    @JvmStatic
    fun extractMonoMod(context: Context): Boolean {
        val targetDir = getMonoModInstallPath()
        Timber.i("正在解压 MonoMod 到 $targetDir")

        return try {
            TemporaryFileAcquirer().use { tfa ->
                Files.createDirectories(targetDir)
                val tempZip = tfa.acquireTempFilePath("monomod.zip")

                context.assets.open(ASSETS_MONOMOD_ZIP).use { input ->
                    Files.copy(input, tempZip, StandardCopyOption.REPLACE_EXISTING)
                }

                when (val result = ArchiveExtractor.builder()
                    .from(tempZip)
                    .to(targetDir)
                    .callback { event ->
                        when (event) {
                            is ArchiveExtractor.Event.Progress -> {
                                Timber.d("解压中: ${event.message} (${(event.progress * 100).toInt()}%)")
                            }
                            is ArchiveExtractor.Event.Complete -> Timber.i("MonoMod 解压完成")
                            is ArchiveExtractor.Event.Error -> {
                                Timber.e(event.cause, "解压错误: ${event.message}")
                            }
                        }
                    }
                    .build()
                    .extract()
                ) {
                    is ArchiveExtractor.Result.Success -> Unit
                    is ArchiveExtractor.Result.Failure -> return false
                }

                Timber.i("MonoMod 已解压到 $targetDir")
                true
            }
        } catch (e: Exception) {
            Timber.e(e, "解压 MonoMod 失败")
            false
        }
    }

    @JvmStatic
    fun applyMonoModPatches(context: Context, gameDirectory: String): Int {
        return applyMonoModPatches(context, gameDirectory, true)
    }

    @JvmStatic
    fun applyMonoModPatches(context: Context, gameDirectory: String, verboseLog: Boolean): Int {
        return try {
            val patchAssemblies = loadPatchArchive(context)
            if (patchAssemblies.isEmpty()) {
                if (verboseLog) Timber.w("MonoMod 目录为空或不存在")
                return 0
            }

            val gameDir = File(gameDirectory)
            val gameAssemblies = findGameAssemblies(gameDir)

            var patchedCount = 0
            for (assemblyFile in gameAssemblies) {
                val assemblyName = assemblyFile.name
                patchAssemblies[assemblyName]?.let { data ->
                    if (replaceAssembly(assemblyFile, data)) {
                        if (verboseLog) Timber.d("已替换: $assemblyName")
                        patchedCount++
                    }
                }
            }

            if (verboseLog) Timber.i("已应用 MonoMod 补丁，替换了 $patchedCount 个文件")
            patchedCount
        } catch (e: Exception) {
            Timber.e(e, "应用补丁失败")
            -1
        }
    }

    private fun loadPatchArchive(context: Context): Map<String, ByteArray> {
        val assemblies = mutableMapOf<String, ByteArray>()
        try {
            val monoModPath = getMonoModInstallPath()
            val monoModDir = monoModPath.toFile()

            if (!monoModDir.exists() || !monoModDir.isDirectory) {
                Timber.w("MonoMod 目录不存在: $monoModPath")
                return assemblies
            }

            val dllFiles = findDllFiles(monoModDir)
            Timber.d("从 $monoModPath 找到 ${dllFiles.size} 个 DLL 文件")

            for (dllFile in dllFiles) {
                try {
                    val assemblyData = Files.readAllBytes(dllFile.toPath())
                    assemblies[dllFile.name] = assemblyData
                } catch (e: Exception) {
                    Timber.w(e, "读取 DLL 失败: ${dllFile.name}")
                }
            }
        } catch (e: Exception) {
            Timber.e(e, "加载 MonoMod 补丁失败")
        }
        return assemblies
    }

    private fun findDllFiles(directory: File): List<File> {
        val dllFiles = mutableListOf<File>()
        directory.listFiles()?.forEach { file ->
            if (file.isDirectory) {
                dllFiles.addAll(findDllFiles(file))
            } else if (file.name.endsWith(".dll")) {
                dllFiles.add(file)
            }
        }
        return dllFiles
    }

    private fun findGameAssemblies(directory: File): List<File> {
        if (!directory.exists() || !directory.isDirectory) return emptyList()

        val assemblies = mutableListOf<File>()
        directory.listFiles()?.forEach { file ->
            if (file.isDirectory) {
                assemblies.addAll(findGameAssemblies(file))
            } else if (file.name.endsWith(".dll")) {
                assemblies.add(file)
            }
        }
        return assemblies
    }

    private fun replaceAssembly(targetFile: File, assemblyData: ByteArray): Boolean {
        return try {
            FileOutputStream(targetFile).use { it.write(assemblyData) }
            true
        } catch (e: Exception) {
            Timber.e(e, "替换失败: ${targetFile.name}")
            false
        }
    }
}
